ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionPool.java

```diff
@@ -380,8 +381,14 @@ public void onBlockAdded(final BlockAddedEvent event) {
 
   private void reAddTransactions(final List<Transaction> reAddTransactions) {
     if (!reAddTransactions.isEmpty()) {
-      var txsByOrigin =
+      // if adding a blob tx, and it is missing its blob, is a re-org and we should restore the blob
+      // from cache.
+      List<Transaction> restoredTransactions =
           reAddTransactions.stream()
+              .map(t -> pendingTransactions.restoreBlob(t).orElse(t))
+              .toList();
+      var txsByOrigin =
+          restoredTransactions.stream()
```

Comment:
Opt: concatenate the streams

```suggestion
      List<Transaction> txsByOrigin =
          reAddTransactions.stream()
              .map(t -> pendingTransactions.restoreBlob(t).orElse(t))
```